### ISSUE TITLE
haskell serialized formats could be better supported

### ISSUE THREAD

**Issue by oli-obk** [serde-rs/serde#758]
haskell serializes enums to json by encoding a `t` field for the variant name and a `c` field for the content. We can represent this with

```rust
#[derive(Serialize, Deserialize)]
#[serde(tag = "t")]
enum MyType {
    Variant { c: ActualContent },
    Variant2 { c: (A, B, C) },
}
```

This works fine, but is a little verbose. Not sure if this use case is too niche.

**Comment by dtolnay** [comment]
This representation also came up in https://github.com/serde-rs/serde/issues/415#issue-162867209 so I don't think it is too niche. How about something like this?

```rust
#[derive(Serialize, Deserialize)]
#[serde(tag = "t", content = "c")]
enum MyType {
    Variant(ActualContent),
    Variant2(A, B, C),
}
```

**Comment by oli-obk** [comment]
Sounds simple enough. We need to be aware of edge cases though. like what happens with empty variants (don't serialize the content field, deserialize without content field and with "unit" content field).

**Comment by elliottslaughter** [comment]
I took a first pass on this in #771, but got stuck at the part where I need to serialize the internal contents of the `"c": ...` portion. If you have any advice, I'd appreciate suggestions in the other thread.

**Comment by dtolnay** [comment]
Serialization has been implemented in #771. I would like an implementation of deserialization as well before publishing this.

**Comment by dtolnay** [comment]
I released this in [0.9.8](https://github.com/serde-rs/serde/releases/tag/v0.9.8).
